core/src/main/java/io/aiven/kafka/tieredstorage/metrics/MeteredInputStream.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.metrics;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.function.Consumer;
+
+import org.apache.kafka.common.utils.Time;
+
+public class MeteredInputStream extends InputStream {
```

Comment:
There's a problem with this implementation: direct inheritance from `InputStream` and overriding only `read()` will result in other read methods being implemented through `read()`, which is very inefficient, because it's by-byte.

I looked at this problem some time ago and here's my draft: https://github.com/Aiven-Open/tiered-storage-for-apache-kafka/tree/ivanyu/time-to-first-byte. Using `FilterInputStream` allows to override fewer methods.

Reply:
Thanks for the pointer! Yes, at the end other implementations use the read method, but as you mentioned it could be inefficient. Will revisit this then.